# 错误的类型

前端主要有以下四种错误类型： 

- 未捕获的错误
- 未捕获的Promise错误
- 网络请求错误
- 资源加载错误

## 未捕获的错误

什么是未捕获的错误呢？举个例子

```js
const a = {};
a.b.c; // Uncaught TypeError: Cannot read properties of undefined (reading 'c')
console.log(a.b);
```

像上述这种尝试从 `undefined` 或则 `null` 对象上读取值的行为，叫做类型错误(TypeError)，它是 `javascript` 在运行过程中发生的错误，被叫做**运行时错误**。

不仅如此，上面的错误还是一个**同步错误**，也就是 `a.b.c` 这个报错是同步发生的，会导致下面的 `console` 语句无法执行。那么相对的，我们有**异步错误**的概念

```js
const a = {};
setTimeout(() => {
  a.b.c; // Uncaught TypeError: Cannot read properties of undefined (reading 'c')
}, 0)
console.log(a.b);
```

这个例子中错误就是**异步错误**，它的发生在 `setTimeout` 或者 `setInterval` 等异步函数中，不会阻塞主流程的代码执行。

**异步错误**很有用。比如，我们想做这么一个事情： 在后端接口返回的商品少于两个数量的时候，进行一次自定义错误的上报。这个时候我们就可以使用异步错误来进行手动上报。

```tsx
const throwErrorAysnc = (error) => {
  setTimeout(() => {
    throw error
  }, 0)
}

if(items.length < 2){
  throwErrorAysnc(new Error("商品数量少于两个"));
}
```

当然，每个事物都有其双面性。异步错误的存在并不会影响主线程代码的执行，然而，由于浏览器的事件循环机制，异步代码的错误堆栈信息可能会相对较少。这是因为，**异步任务需要等待所有同步任务完成后才能从异步队列中取出并执行**。在这个过程中，我们无法通过调用栈回溯到异步任务创建时的原始堆栈信息。

## 未捕获的Promise错误

未捕获的 `Promise` 错误和上述的错误有什么不同呢？ 你是否好奇为什么要把 `Promise` 单独拿出来讲？

```js
const promise = new Promise((resolve, reject) => {
  throw new Error('This is a promise error');
});

promise.catch(error => {
  console.log('Caught a promise error: ');
});
```

来看看这个例子，我们明明执行了 `throw new Error` 这个操作，但是控制台却并没有抛出错误。这是因为从 `Promise` 的语义来说，`Promise` 中抛出的错误不需要直接处理。相反，这个错误会导致 `Promise` 进入rejected状态。。然后，如果在一段时间后（通常是微任务队列清空后），这个rejected状态的 `Promise` 还没有被处理，那么浏览器会触发 window 对象的 unhandledrejection 事件，并可能在控制台打印一个警告。

```tsx
window.addEventListener('unhandledrejection',error => {
  console.log('error',error)
})

const promise = new Promise((resolve, reject) => {
  throw new Error('This is a promise error');
});
```

这个例子中，我们就成功地在控制台打印了 `Promise` 错误。

所以，浏览器区分了未捕捉的异常和未捕捉的 `Promise` 异常，这是因为 `Promise` 对象有特殊的错误处理机制

## 网络请求错误

在进行网络请求时，可能会遇到各种错误。这些错误可能源于客户端、服务器或网络本身。我们可以将这些错误大致归为以下几类：

客户端错误：这类错误通常表示为4xx HTTP状态码。例如，如果请求的资源不存在，服务器可能会返回404 Not Found错误。如果客户端发送的请求数据格式不正确，服务器可能会返回400 Bad Request错误。

服务器错误：这类错误通常表示为5xx HTTP状态码。例如，如果服务器内部发生错误，可能会返回500 Internal Server Error。如果服务器正在进行维护或过载，可能会返回503 Service Unavailable。

网络错误：这类错误可能是由于网络连接问题导致的，如DNS解析失败、超时、断网等。

```tsx
fetch('/api/data')
  .then(response => {
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return response.json();
  })
  .then(data => console.log(data))
  .catch(error => console.error('An error occurred: ', error));
```

## 资源加载错误

资源错误是非常常见的一类错误。资源错误通常指的是在加载静态资源（如JavaScript文件、CSS文件、图片等）时发生的错误。这可能是由于资源文件不存在、服务器错误、网络连接问题、浏览器兼容性问题等原因导致的。

例如，当一个<img>标签试图加载一个不存在的图片URL时，浏览器会触发一个错误事件。

```tsx
<img src="http://rqwrqwerq.com/index.png"/>
```